cli/fathom_web/commands/test.py

```diff
@@ -0,0 +1,77 @@
+from json import JSONDecodeError, load, loads
+
+from click import argument, BadParameter, command, File, style
+
+from ..accuracy import accuracy_per_tag, accuracy_per_page, pretty_accuracy
+from ..utils import classifier, tensor, tensors_from
+
+
+def decode_weights(ctx, param, value):
+    """Validate a click option, making sure it's a valid JSON object with
+    properly formatted "coeff" and "bias" keys."""
+    try:
+        decoded_weights = loads(value)
+    except JSONDecodeError:
+        raise BadParameter('Weights must be a valid JSON object.')
+    else:
+        if 'coeffs' in decoded_weights and 'bias' in decoded_weights:
```

Comment:
You can avoid this amount of nesting by checking for the bad situations and raising exceptions.
```
if 'coeffs' not in decoded_weights and 'bias' not in decoded_weights:
    raise BadParameter('Weights must contain "coeffs" and "bias" keys.')
if not isinstance(decoded_weights['bias'], float):
    raise BadParameter('Bias must be a float.')
if not isinstance(decoded_weights['coeffs'], list) or any((len(pair) != 2 or not isinstance(pair[0], str) or not isinstance(pair[1], float)) for pair in decoded_weights['coeffs']):
    raise BadParameter('Coeffs must be a list of 2-element lists: [["ruleName", numericCoefficient], ...].')
return decoded_weights
```

Reply:
There is a bit of personal preference here, I think. I tend to not mind nesting or width in general and instead dislike early exits, because they make me read not only the whitespace but also scan for about 5 different keywords (break, continue, return, raise, etc.). But I'm fine flattening this out if you prefer, as the structure is nice and regular. I mostly used your code, but I flopped the final nots with De Morgan's law to avoid having quite so many negations, as those exceed my brain's capacity. :-) Patch to follow.